src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -531,11 +587,22 @@ class GroupReadsByUmi
 
   /** Consumes the next group of templates with all matching end positions and returns them. */
   def takeNextGroup(iterator: BufferedIterator[Template]) : Seq[Template] = {
-    val first = iterator.next()
+    val first     = iterator.next()
     val firstEnds = ReadInfo(first.r1.getOrElse(fail(s"R1 missing for template ${first.name}")))
-    val buffer = ListBuffer[Template]()
-    while (iterator.hasNext && firstEnds == ReadInfo(iterator.head.r1.get)) buffer += iterator.next()
-    first :: buffer.toList
+    val firstUmi  = first.r1.get.apply[String](this.assignTag)
+    val builder   = IndexedSeq.newBuilder[Template]
+    builder    += first
+
+    while (
+      iterator.hasNext &&
+      firstEnds == ReadInfo(iterator.head.r1.get) &&
+      // This last condition only works because we put a canonicalized UMI into rec(assignTag) if canTakeNextGroupByUmi
+      (!canTakeNextGroupByUmi || firstUmi == iterator.head.r1.get.apply[String](this.assignTag))
```

Comment:
So this does change behavior in a subtle way.  Suppose we have three templates that have all but the same assign tag.  Let's say `AAAA`, `GGGGG`, `GGGGT`, with `--min-umi-length=3`.  

Previously, all templates would be read into memory, and `truncateUmis` would truncate to the length of smallest UMI observed in the group of templates, in this case 4bp long due to the `AAAA` (not 3 as per the command line!).  So the three templates would have UMIs `AAAA->AAAA`, `GGGGG->GGGG`, and `GGGGT->GGGG`.  So we'd assign two unique molecules (`AAAA` and `GGGG`, with the last molecule containing the last two reads).

In the new implementation, we truncate the raw UMI bases based on `--min-umi-length=3` to set `MI` for sorting.  So we'd truncate to length 3 for sorting: `AAAA->AAA`, `GGG->GGG`, and `GGG->GGG`.  When we read back in after sorting, we read in the first template by itself (only read with `MI:AAA`), so no truncation is applied and it stays the same (`AAAA`).  We then read in all templates with MI having `GGG`, which gives the second two templates.  Now we go back to the raw UMIs to find the length of the shortest UMI of the two.  Both are 5bp long, so we do not truncate and keep them the same (`GGGGG` and `GGGGT`).  But now these UMIs differ, so we get three molecules!

One _could_ argue that the new implementation is an improvement, but it does change behavior subtly.

Reply:
Hrm, that is an interesting point.  So basically if you have variable length single UMIs _and_ have edits = 0 the behavior will be subtly different.  My instinct is to call this an improvement and move on, but I don't have a great sense of who (or on what kind of data) the variable length support is used, so I'm not 100% sure.